pkg/codegen/pcl/invoke.go

```diff
@@ -44,6 +44,56 @@ func getInvokeToken(call *hclsyntax.FunctionCallExpr) (string, hcl.Range, bool)
 	return literal.Val.AsString(), call.Args[0].Range(), true
 }
 
+// annotateObjectProperties annotates the properties of an object expression with the types of the corresponding
+// properties in the schema. This is used to provide type information
+// for invoke calls that didn't have type annotations.
+//
+// this function will recursively annotate the properties of objects
+// that are nested within the object expression type.
+func annotateObjectProperties(modelType model.Type, schemaType schema.Type, topLevelAnnotate bool) {
+	if optionalType, ok := schemaType.(*schema.OptionalType); ok {
+		schemaType = optionalType.ElementType
+	}
+
+	if objectType, ok := modelType.(*model.ObjectType); ok {
+		if schemaObjectType, ok := schemaType.(*schema.ObjectType); ok {
+			schemaProperties := make(map[string]schema.Type)
+			for _, schemaProperty := range schemaObjectType.Properties {
+				schemaProperties[schemaProperty.Name] = schemaProperty.Type
+			}
+			// top-level annotation, only when needed
+			if topLevelAnnotate {
+				objectType.Annotations = append(objectType.Annotations, schemaType)
+			}
+
+			// now for each property, annotate it with the associated type from the schema
+			for propertyName, propertyType := range objectType.Properties {
+				if associatedType, ok := schemaProperties[propertyName]; ok {
+					annotateObjectProperties(propertyType, associatedType, true)
+				}
+			}
+		}
+	}
+
+	if arrayType, ok := modelType.(*model.ListType); ok {
+		underlyingArrayType := arrayType.ElementType
+		if schemaArrayType, ok := schemaType.(*schema.ArrayType); ok {
+			underlyingSchemaArrayType := schemaArrayType.ElementType
+			annotateObjectProperties(underlyingArrayType, underlyingSchemaArrayType, true)
+		}
+	}
+
+	if tupleType, ok := modelType.(*model.TupleType); ok {
+		if schemaArrayType, ok := schemaType.(*schema.ArrayType); ok {
+			underlyingSchemaArrayType := schemaArrayType.ElementType
+			elementTypes := tupleType.ElementTypes
+			for _, elemType := range elementTypes {
+				annotateObjectProperties(elemType, underlyingSchemaArrayType, true)
+			}
+		}
+	}
```

Comment:
Shouldn't we handle `model.Union` types here?

Reply:
I've added support to it like this: 
```go
	case *model.UnionType:
		if schemaUnionType, ok := schemaType.(*schema.UnionType); ok {
			modelElementTypes := arg.ElementTypes
			schemaElementTypes := schemaUnionType.ElementTypes
			if len(modelElementTypes) == len(schemaElementTypes) {
				// annotate each type inside the model union with its corresponding
				// schema type from the schema union.
				// Assuming here that they match based on the index (ordered), is this correct?
				for i := range modelElementTypes {
					annotateObjectProperties(modelElementTypes[i], schemaElementTypes[i])
				}
			}
		}
	}
```
but I am not sure my assumption is correct here that schema union element types and model union element types are ordered so that their indexes match. Thoughts?

Reply:
I'm not aware of any guarantee of ordering between model union elements and schema union elements. Keep in mind that `*model.Union` is not isomorphic to `*schema.Union`. For example `schema.Option{Element: T}` is expressed as `model.Union{Elements: [None, T]}`. If nothing else, this needs to be handled during the optional check at the beginning of `annotateObjectProperties`. My understanding is that we are allowed to squash model unions, so that `Union[foo, Union[None, bar]]` should be equivalent to `Union[foo, None, bar]`. Schema unions are more limited, since they need to be translatable into `go`, which doesn't have union types.

Reply:
@iwahbe Alright, I'll make a TODO issue out of handling unions and tackle it later with a specific test schema for these cases because I really haven't seen any invokes that have unions as input but I would like to get this fix merged and use it in #10976

Reply:
In that case I would change the union handling to collapse `Union[None, T]` to `T` and punt on the rest. If we aren't going to address unions here, we shouldn't annotate when we are unsure.

Reply:
@iwahbe I've implemented it like this:
```go
case *model.UnionType:
	// sometimes optional schema types are represented as unions: None | T
	// in this case, we want to collapse the union and annotate the underlying type 
	if len(arg.ElementTypes) == 2 && arg.ElementTypes[0] == model.NoneType {
		annotateObjectProperties(arg.ElementTypes[1], schemaType)
	} else if len(arg.ElementTypes) == 2 && arg.ElementTypes[1] == model.NoneType {
		annotateObjectProperties(arg.ElementTypes[0], schemaType)
	} else {
		// TODO https://github.com/pulumi/pulumi/issues/10993
		// We need to handle the case where the schema type is a union type.
	}
}
```